### ISSUE TITLE
SC2093 false positive (when exec used for redirection)

### ISSUE THREAD

**Issue by sqweek** [koalaman/shellcheck#1180]
This looks similar to #697 except I can still reproduce it online.

- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash
exec {origout}>&1- >"tmp.log" 2>&1

echo $origout

pwd
```

#### Here's what shellcheck currently says:

```
Line 2:
exec {origout}>&1- >"tmp.log" 2>&1
^-- SC2093: Remove "exec " if script should continue after this command.
```

Interestingly if you comment out the `pwd` or replace it with an `echo` (even `echo $(pwd)`) then the SC2093 warning disappears. Somehow it depends on the commands coming after the exec...

#### Here's what I wanted or expected to see:

No output.

**Comment by charlescanato** [comment]
But, considering that the checker can't know what's in `origout`, I feel that the warning would be valid (if `origout` could contain a command). But I agree it should be consistent, and not appear only based on the next commands.

**Comment by sqweek** [comment]
The contents of `origout` don't matter. The `{variable}>&1-` redirection *assigns* the allocated FD to the variable, no expansion takes place. From the man page:

>       Each redirection that may be preceded by a file descriptor number may instead be
>       preceded by a word of the form {varname}.  In this case,  for  each  redirection
>       operator  except  >&- and <&-, the shell will allocate a file descriptor greater
>       than or equal to 10 and assign it to varname.  If >&-  or  <&-  is  preceded  by
>       {varname}, the value of varname defines the file descriptor to close.

Note that the `>&-` operator mentioned here is different to the `>&DIGIT-` operator I'm using so the exception doesn't apply (see also "Moving File Descriptors" in the man page).

**Comment by koalaman** [comment]
Nice catch! 

The parser didn't recognize the FD move operations and stopped parsing it after `{origout}>&1`. The `-` then became a regular shell word that was interpreted as a command name, which is why ShellCheck thought it would stop executing.

This has now been fixed, thanks!
